bpo-44991: Make GIL handling more explicit in `sqlite3` callbacks

erlend-aasland opened the pull request:
- acquire the GIL at the very start[1]
- release the GIL at the very end

[1] The trace callback performs a sanity check before acquiring the GIL

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-44991](https://bugs.python.org/issue44991) -->
https://bugs.python.org/issue44991
<!-- /issue-number -->

Automerge-Triggered-By: GH:encukou

Commit:
Acquire GIL at the very start of callbacks; release at the very end

Exception: in the trace callback, do a sanity check before attempting to
hold the GIL.

erlend-aasland commented on Modules/_sqlite/connection.c:
```diff
@@ -1461,13 +1451,13 @@ pysqlite_collation_callback(
     int result = 0;
-    gilstate = PyGILState_Ensure();
 
     if (PyErr_Occurred()) {
```
Digression: is `PyErr_Occurred()` really needed? If `PyGILState_Ensure()` fails, won't we just crash intensely anyway?

encukou replied:
I assume this checks for an error set *before this callback is called*, rather than an error in `PyGILState_Ensure`.
`PyErr_Occurred()` (rather than the return value) tells us if an error occured, because the collation callback can't use a return value to signalize an error.
[PyLong_AsLong](https://docs.python.org/3/c-api/long.html#c.PyLong_AsLong) has somewhat similar behavior.

erlend-aasland replied:
I'm not sure I follow you. `pysqlite_collation_callback()` is called from within SQLite; IOW by the SQLite C API `sqlite3_step()`.

Here's a backtrace, showing the last part of the call hierarchy for `sqlite3.Cursor.execute(some_sql_code)`:
```
  * frame #0: 0x0000000105575aab _sqlite3.cpython-311d-darwin.so`pysqlite_collation_callback(context=0x0000000101596080, text1_length=1, text1_data=0x000000010307962b, text2_length=1, text2_data=0x0000000103079643) at connection.c:1463:1
    frame #1: 0x000000010560a2a5 libsqlite3.0.dylib`sqlite3VdbeRecordCompareWithSkip + 859
    frame #2: 0x000000010561096e libsqlite3.0.dylib`vdbeSorterMerge + 77
    frame #3: 0x0000000105610717 libsqlite3.0.dylib`vdbeSorterSort + 192
    frame #4: 0x0000000105601f52 libsqlite3.0.dylib`sqlite3VdbeExec + 18582
    frame #5: 0x00000001055dacbb libsqlite3.0.dylib`sqlite3_step + 394
    frame #6: 0x00000001055852b2 _sqlite3.cpython-311d-darwin.so`pysqlite_step(statement=0x00000001180ebb80) at util.c:33:10
    frame #7: 0x000000010557bbf5 _sqlite3.cpython-311d-darwin.so`_pysqlite_query_execute(self=0x000000010690c200, multiple=0, operation=0x0000000106116d40, second_argument=0x0000000101504250) at cursor.c:588:14
    frame #8: 0x000000010557d20e _sqlite3.cpython-311d-darwin.so`pysqlite_cursor_execute_impl(self=0x000000010690c200, sql=0x0000000106116d40, parameters=0x0000000000000000) at cursor.c:697:12
    frame #9: 0x000000010557aee4 _sqlite3.cpython-311d-darwin.so`pysqlite_cursor_execute(self=0x000000010690c200, args=0x00007ffeefbc88a8, nargs=1) at cursor.c.h:69:20
    frame #10: 0x0000000100157ae4 python.exe`method_vectorcall_FASTCALL(func=0x00000001054e7dd0, args=0x00007ffeefbc88a0, nargsf=2, kwnames=0x0000000000000000) at descrobject.c:388:24
    frame #11: 0x000000010013fb91 python.exe`_PyObject_VectorcallTstate(tstate=0x0000000101405f40, callable=0x00000001054e7dd0, args=0x00007ffeefbc88a0, nargsf=2, kwnames=0x0000000000000000) at abstract.h:114:11
    frame #12: 0x00000001001401df python.exe`object_vacall(tstate=0x0000000101405f40, base=0x000000010690c200, callable=0x00000001054e7dd0, vargs=0x00007ffeefbc8aa0) at call.c:734:14
    frame #13: 0x000000010014050b python.exe`_PyObject_CallMethodIdObjArgs(obj=0x000000010690c200, name=0x000000010558ee50) at call.c:825:24
```

Anyway, `pysqlite_collation_callback()` has been like this from its very start when `pysqlite` still was an external project, so there's no hints in the git log.

encukou replied:
Right, but isn't it called multiple times when returning a result?

erlend-aasland replied:
~For a SELECT query with a COLLATE operator, `pysqlite_collation_callback()` will be called once for every row. IOW, it is called once for every `sqlite3_step()`, since, in SQLite, we step through each row one at the time. So, when you iterate through the results of a query, it is called once for every call to `_pysqlite_fetch_one_row()`.~

Sorry, too early in the morning. It is of course called multiple times from within the first `sqlite3_step` in order to filter the query! So yes, it is called multiple times. This goes for most if not all UDF's.

erlend-aasland replied:
This code snippet...
```
# left out: creating connection 'cx' and collation 'mycoll'
with cx:
    cx.execute("create table t(x, y)")
    cx.executemany("insert into t(x, y) values(?, ?)",
                   [("a", "a"), ("b", "b")])
cx.execute("select x, y from t where x = y collate mycoll")
```

... generates the following calls (simplified):
```
sqlite3_step(sql=create table t(x, y))
sqlite3_step(sql=insert into t(x, y) values(?, ?))
sqlite3_step(sql=insert into t(x, y) values(?, ?))
sqlite3_step(sql=COMMIT)
sqlite3_step(sql=select x, y from t where x = y collate mycoll)
pysqlite_collation_callback(ctx=0x1075b4a30, "a", "a")
pysqlite_collation_callback(ctx=0x1075b4a30, "b", "b")
```

erlend-aasland replied:
~Anyway, don't let this digression stand in the way of this PR :)~ Actually, maybe we should.

You were right in your assumption. I suggest we add a comment about it, and maybe even move this check to the very start of the function. Something like this:

```c
static int
pysqlite_collation_callback(
        void* context,
        int text1_length, const void* text1_data,
        int text2_length, const void* text2_data)
{   
    /* This callback may be executed multiple times per sqlite3_step(). Bail if
     * the previous call failed */
    if (PyErr_Occurred()) {
        return NULL;
    }

    PyGILState_STATE gilstate = PyGILState_Ensure();

    // the rest of the function follows...
```

encukou replied:
`PyErr_Occurred` is Python API and must be called with the GIL held.
The comment is a good idea though :)

erlend-aasland replied:
> PyErr_Occurred is Python API and must be called with the GIL held.

Doh, of course 🤦🏻 😄  I'll add a comment right away.

erlend-aasland replied:
See 2a2152dd40bbf78c4d160a98720b264c1fc60cdf

Commit:
Add explanatory comment about mysterious PyErr_Occurred()

Commit:
Address review

- since _destructor is a callback, wrap it in GIL lock/unlock
- remove GIL lock/unlock from create_callback_context()

encukou reviewed (approved):
Looks good, thank you!
